### PR TITLE
Update to coffeescript 2 and ember-cli 3.0

### LINKED ISSUE #112: project build freeze in v1.14.0
With new version of ember-cli-coffeescript (1.14.0) when I rewrite route.js file into coffee and then I tried to build app, building process freezes or was in infinite loop. No error or another message it just keeps building. I tried that on new clean project on ember 2.4.2 just with ember-cli-coffescript installed. I switched to older version ember-cli-coffescript@1.13.2 and now it works pretty good. So I guess there is some bug in new version.

### LINKED ISSUE #144: Support coffeescript 2
Hi there,

Coffeescript 2 is getting real: http://coffeescript.org/v2/ (https://github.com/coffeescript6/discuss/issues/80).

It supports many **ES2015 features**. The reason why I would love to see it in ember is the **async/await** and **generator** support. The most important reason though is that the nasty import/export hacks using backticks to write plain JS is not needed any more:

```
export default Ember.Route.extend

  model: ->
    @foo()

  foo: ->
    a = await @bar()
    b = await @biz()
    "#{a}-#{b}"

  bar: ->
    new Ember.RSVP.Promise (resolve) ->
      setTimeout ->
        resolve("baz")
      , 2000

  biz: ->
    new Ember.RSVP.Promise (resolve) ->
      setTimeout ->
        resolve("biz")
      , 2000
```

I've made a quick test and it is simple to get it running by altering only a few lines of ember-cli-coffeescript and [broccoli-coffee](https://github.com/joliss/broccoli-coffee):

- update the package.json, remove coffee-script, add coffeescript
- update every require('coffee-script') to require('coffeescript')

But I think this needs more discussion on a meta level, as Coffeescript 1 runs in every browser, but Coffeescript 2 needs transpilation. Thankfully Ember includes Babel, so that is also pretty easy to accomplish:

```
var app = new EmberApp(defaults, {
    'ember-cli-babel': {
      includePolyfill: true
    }
  });
```

That's the reason why I did not just create pull requests. Maybe new repos should be created, called ember-cli-coffeescript-2 and babel-coffee-2? That would ensure that there is a clear distinction between old Coffeescript (1) and ES2015 Coffeescript (2).

Looking forward to hear your opinions.

BR, toovy

### PR DESCRIPTION
**Pull request by jakesjews**
Resolves #144 

This should also resolve #112 since https://github.com/joliss/broccoli-coffee/issues/20 was resolved in the latest broccoli-coffee

### CODE REVIEW

File: lib/utilities/string.js [hunk 1]
```diff
@@ -4,9 +4,9 @@
 var STRING_DASHERIZE_REGEXP = (/[ _]/g);
 var STRING_DASHERIZE_CACHE = {};
 var STRING_DECAMELIZE_REGEXP = (/([a-z\d])([A-Z])/g);
-var STRING_CAMELIZE_REGEXP = (/(\-|_|\.|\s)+(.)?/g);
+var STRING_CAMELIZE_REGEXP = (/(-|_|\.|\s)+(.)?/g);
```
Comment (jakesjews) [inline lib/utilities/string.js]: eslint flagged some unneeded escape characters in these regexes

**Review by auvipy — APPROVED** [review]

**Comment by jakesjews** [comment]
Is there anything else that needs to be done to get this merged?

**Comment by kimroen** [comment]
@jakesjews Thank you so much for looking at this!

I was wondering: How do these changes affect compatibility? Could I still install `ember-cli-coffeescript` on my project using Ember 1.13 after this has been released?

Does this change required Node versions for the project?

I merged some changes, so there are some conflicts that need resolving - sorry about that.

**Comment by jakesjews** [comment]
@kimroen merged up now. Coffeescript 2.0 requires node version 6 or greater. I'm not quite sure if it works on ember 1.13 or not but I think it has reached EOL.

**Comment by mriska** [comment]
It would be awesome to get this PR merged. I suggest making this version `2.0.0` of ember-cli-coffeescript, so updating would be a deliberate choice for existing projects.

**Comment by boris-petrov** [comment]
@kimroen - ping. :)

**Comment by nudded** [comment]
@kimroen re-ping, This has now become urgent as the new Ember testing API's use async everywhere

**Comment by youroff** [comment]
+1 and thanks, @jakesjews, great job! Hope it's merged soon, we also depend on coffeescript and use async api.

**Comment by jakesjews** [comment]
If anyone wants to use it I have published https://www.npmjs.com/package/eflex-ember-cli-coffeescript which uses coffeescript 2.0. The company I work for has been using it in production for a long time now.

**Comment by youroff** [comment]
That's awesome! We were using your branch in dev, but now that it's published as a separate package, will definitely switch to it.

**Comment by kimroen** [comment]
@jakesjews Would you be interested in helping maintain this package? The main reason I haven't done anything with this is that I have bigger plans for what I want to do with it, but I just never seem to find the time and energy for it.

**Comment by jakesjews** [comment]
@kimroen yeah I'm definitely willing to help

**Comment by ovikholt** [comment]
Any action here? :)

**Pull request closed without merging**